agent/api/task/task.go

```diff
@@ -536,11 +551,37 @@ func (task *Task) addNetworkResourceProvisioningDependency(cfg *config.Config) {
 		if container.IsInternal() {
 			continue
 		}
-		container.BuildContainerDependency(PauseContainerName, apicontainerstatus.ContainerResourcesProvisioned, apicontainerstatus.ContainerPulled)
+		container.BuildContainerDependency(NetworkPauseContainerName, apicontainerstatus.ContainerResourcesProvisioned, apicontainerstatus.ContainerPulled)
 		pauseContainer.BuildContainerDependency(container.Name, apicontainerstatus.ContainerStopped, apicontainerstatus.ContainerStopped)
 	}
 }
 
+func (task *Task) addNamespaceSharingProvisioningDependency(cfg *config.Config) {
+	// Pause container does not need to be created if no namespace sharing will be done
+	if task.getIpcMode() == "" && task.getPidMode() == "" {
+		return
+	}
+
+	// Create a Pause container if either IPC resource or PID namespace needs to be shared
+	if task.getIpcMode() == "task" || task.getPidMode() == "task" {
+		namespacePauseContainer := apicontainer.NewContainerWithSteadyState(apicontainerstatus.ContainerResourcesProvisioned)
+		namespacePauseContainer.TransitionDependenciesMap = make(map[apicontainerstatus.ContainerStatus]apicontainer.TransitionDependencySet)
+		namespacePauseContainer.Name = NamespacePauseContainerName
+		namespacePauseContainer.Image = fmt.Sprintf("%s:%s", cfg.PauseContainerImageName, cfg.PauseContainerTag)
```

Comment:
To avoid `PauseContainerImageName` being overridden by Fargate, one option is you can use `DefaultPauseContainerImageName` here. And you will also need to make change [here](https://github.com/aws/amazon-ecs-agent/blob/e8b5229a9ea8b748ebad4db4c098ccc8df2dcd5f/agent/app/agent_unix.go#L84) and `pause_linux.go` to make Agent always load *default* pause container tarball